docs/docs/reference/dropped-features/package-objects.md

```diff
@@ -0,0 +1,44 @@
+---
+layout: doc-page
+title: "Dropped: Package Objects"
+---
+
+Package objects
+```scala
+    package object p {
+      val a = ...
+      def b = ...
+    }
+```
+will be dropped. They are still available in Scala 3.0, but will be deprecated and removed afterwards.
+
+Package objects are no longer needed since all kinds of definitions can now be written at the top-level. E.g.
+```scala
+    package p
+    type Labelled[T] = (String, T)
+    val a: Labelled[Int] = ("count", 1)
+    def b = a._2
+
+    case class C()
+
+    implicit object Cops {
+      def (x: C) pair (y: C) = (x, y)
+    }
+```
+There may be several source files in a package containing such toplevel definitions, and source files can freely mix toplevel value, method, and type definitions with classes and objects.
+
+The compiler generates synthetic objects that wrap toplevel definitions falling into one of the following categories:
+
+ - all pattern, value, method, and type definitions,
+ - implicit classes and objects,
+ - companion objects of opaque types.
+
+If a source file `src.scala` contains such toplevel definitions, they will be put in a synthetic object named `src$package`. The wrapping is transparent, however. The definitions in `src` can still be accessed as members of the enclosing package.
+
+**Note 1:** This means that the name of a source file containing wrapped toplevel definitions is relevant for binary compatibility. If the name changes, so does the name of the generated object and its class.
+
+**Note 2:** A toplevel main method `def main(args: Array[String]): Unit = ...` is wrapped as any other method. If it appears
+in a source file `src.scala`, it could be invoked from the command line using a command like `scala src$package`. Since the
+"program name" is mangled it is recommended to always put `main` methods in explicitly named objects.
```

Comment:
Is this intentional forever? It seems like it could add a lot of value to have a simpler way of writing programs without the need for the excess baggage of a wrapper object. Similar to @lihaoyi-databricks's suggestion of script files, but supporting programs with arguments.

Reply:
Agreed, but we'd need to come up with a way to do that which fits with everything else, suggestions welcome.

Reply:
> supporting programs with arguments.

The way it's done in Python is just making the program command-line arguments a global variable.

If you want to be fancy, Ammonite and Mill have `@main` methods with auto deserialization of flags into method params, `--help`-text generation, nice error messages, handling of default arguments, etc.. But I suspect that's too much to embed into the core language runtime.

Reply:
There's another problem, independent of command line arguments. The synthetic name of the wrapper is mangled, i.e. `src$package`. The reason for doing this is that before accessing a member of a package the first time, we need to scan all wrappers for their contents. But we certainly do not want to scan all class files in the package. The naming convention tells us which files to scan. There might be another way to do it, but this one seemed to be the most straightforward.

Reply:
I don't think toplevel definitions are the right way to write main programs anyway. The problem is, you still have to write a `main` method with an array parameter, which is bulky. I prefer our current way of expressing this:
```
object Prog extends App {
  ...
}
```